### ISSUE TITLE
训练的时候不收敛

### ISSUE THREAD

**Issue by geyijun** [SHUCV/digit#1]
按照readme中的做法，先conveer,py ,再solve.py。遇到一个问题
发现loss在最低降到0.6之后，就一直无法收敛。
继续检查，发现训练网络中的反卷积层的学习率为0 ，如下
layer {
  name: "upscore_new"
  type: "Deconvolution"
  param {
    lr_mult: 0
  }
不知这样设置是否属于正常。
尝试1>修改lr_mult: 0.01 继续训练，还是无法收敛。
尝试2>修改整体的学习率base_lr，继续训练，还是无法收敛。
尝试3>修改batch_size ,详细一样，无法收敛。
请教一下作者，上述不收敛的现象可能是什么原因?

**Comment by jy9387** [comment]
Hi! 很高兴看到你的问题。


1. 请问你训练时初始的loss为多少，只凭0.6这个数字不是很好判断是否收敛，可以利用当前得到的模型，在训练集上进行预测，通过观察结果判断是否收敛。

另外，训练loss并非越小越好，模型在测试集上的表现更重要，建议也在测试集上进行相似的预测，判断当前模型的泛化能力。


2. FCN网络中，反卷积层的学习率默认为0，我们采取了这种一般做法，不用担心属于正常情况，详细请参考FCN原始paper及其代码。


3. 作为FCN-based的网络，输入图片的大小不一致，因此请保持batch-size为1。


如果还有什么疑惑，再一起交流讨论^ ^


Jiang Yuan
School of Communication Engineering, Shanghai University, Shanghai, P.R. China


________________________________
发件人: geyijun <notifications@github.com>
发送时间: 2016年7月8日 15:13
收件人: SHUCV/digit
主题: [SHUCV/digit] 训练的时候不收敛 (#1)


按照readme中的做法，先conveer,py ,再solve.py。遇到一个问题
发现loss在最低降到0.6之后，就一直无法收敛。
继续检查，发现训练网络中的反卷积层的学习率为0 ，如下
layer {
name: "upscore_new"
type: "Deconvolution"
param {
lr_mult: 0
}
不知这样设置是否属于正常。
尝试1>修改lr_mult: 0.01 继续训练，还是无法收敛。
尝试2>修改整体的学习率base_lr，继续训练，还是无法收敛。
尝试3>修改batch_size ,详细一样，无法收敛。
请教一下作者，上述不收敛的现象可能是什么原因?

―

**Comment by geyijun** [comment]
谢谢作者的回复。我重新将问题简化并阐述一下，
我当前需求是:
    按照项目中的readme把整个digit项目中的效果可以跑出来。
我详细记录我的操作过程如下:
步骤1: git clone digit 项目到本地。
步骤2: 下载data，解压到digit/data目录下 （unzip digit_data.zip）。

第一阶段:用 【作者训练好的网络】+【训练数据集中的图片】，验证效果。
步骤1>按照作者readme中的说，download pre-trained model 得：fcn11_full_iter_15000.caffemodel.放在digit/examples/fdigit/
步骤2>用python写一个简单脚本进行测试。见脚本my_test.py
步骤3>通过人工校验输出结果的正确性。发现完全不对(几乎95%是背景类，仅底边有一点点的0 和1 分类)。
我的问题:::为啥会这样？？？

说明1：caffe是7月1号安装的github上的版本。
说明2:  上述过程没有任何修改网络参数的动作，仅使用pre-trained model 


===========================测试脚本复制 _begin==================================
#!/usr/bin/env python
# -*- coding:utf-8 -*-  
from __future__ import division
caffe_root = '/opt/caffe/'
import numpy as np
import sys,os
sys.path.insert(0, caffe_root + 'python')
import caffe

#创建网络
net = caffe.Net('./fcn11_deploy.prototxt',
                './fcn11_full_iter_15000.caffemodel',
                caffe.TEST)
#加载测试图像                
image = caffe.io.load_image('../../data/fcn_label_full/fig/imag1_1-1135.jpg')
transpose_image = image.transpose(2,0,1)                                #把通道axis移到前面
transpose_image2 = transpose_image.reshape((1,)+transpose_image.shape)  #增加一个axis匹配data的要求

#把图像拷贝到网络中
net.blobs['data'].reshape(1,         # batch size
                          3,         # 3-channel (BGR) images
                          50, 160)   # image size is 50*160
net.blobs['data'].data[...] = transpose_image2

#执行网络前向运算
output = net.forward() #输出的是最后一层的top blob

#输出预测结果
#对每一个像素点输出其最大预测值对应的通道数
#因为我们知道shape(1,11,50,160)
temp_image = output['score1']
vote_image = np.zeros((50,160),dtype=np.int8) 
for i in range(50):
    for j in range(160):
        max_index = 0
        for ch in range(11):
            if temp_image[0,ch,i,j] >= temp_image[0,max_index,i,j]:
                max_index = ch
                vote_image[i,j] = max_index
np.savetxt("vote_image.txt",vote_image,fmt="%02d")
===========================测试脚本复制 _end==================================

**Comment by jy9387** [comment]
你好，我明白了你的意思，并就你说的问题自己进行了测试。


*如果想按照README自己训练model，还请使用我们的caffe，因为代码跟官方的还是有一些不一样的，主要是对softmax_loss.cpp层进行了一些修改，加入了多分类时正负样本的均衡，从而改善训练过程。之前你说过loss到0.6无法继续下降，有可能是因为这个原因。我对git上的代码做过测试，可以正常编译，而且按照README来训练，应该是可以收敛的。如果这方面还有什么疑问，可以再详细讨论。


*我们提供了matlab版本的测试接口。如果不考虑训练，只使用我们预训练的model进行预测，应该是可以得到正常结果的，建议先用`examples/fdigit/test_fcn11_full.m`测试几张图像看看结果。


*大致看了下你的python脚本，无法得到正确预测结果的主要原因，可能是因为你是对卷积网络输出的所有11层进行取最大的操作，而我们的想法是忽略第11层（负类），只对前10层取最大得到预测结果。可以根据我们上传的matlab版本的测试接口，对你的python脚本进行相应修改。


如果没有明白或是有任何其他疑问，欢迎再交流

________________________________
发件人: geyijun <notifications@github.com>
发送时间: 2016年7月9日 11:30
收件人: SHUCV/digit
抄送: jy9387; Comment
主题: Re: [SHUCV/digit] 训练的时候不收敛 (#1)


谢谢作者的回复。我重新将问题简化并阐述一下，
我当前需求是:
按照项目中的readme把整个digit项目中的效果可以跑出来。
我详细记录我的操作过程如下:
步骤1: git clone digit 项目到本地。
步骤2: 下载data，解压到digit/data目录下 （unzip digit_data.zip）。

第一阶段:用 【作者训练好的网络】+【训练数据集中的图片】，验证效果。
步骤1>按照作者readme中的说，download pre-trained model 得：fcn11_full_iter_15000.caffemodel.放在digit/examples/fdigit/
步骤2>用python写一个简单脚本进行测试。见脚本my_test.py
步骤3>通过人工校验输出结果的正确性。发现完全不对(几乎95%是背景类，仅底边有一点点的0 和1 分类)。
我的问题:::为啥会这样？？？

说明1：caffe是7月1号安装的github上的版本。
说明2: 上述过程没有任何修改网络参数的动作，仅使用pre-trained model

===========================测试脚本复制 _begin==================================
#!/usr/bin/env python

-- coding:utf-8 --

from future import division
caffe_root = '/opt/caffe/'
import numpy as np
import sys,os
sys.path.insert(0, caffe_root + 'python')
import caffe

#创建网络
net = caffe.Net('./fcn11_deploy.prototxt',
'./fcn11_full_iter_15000.caffemodel',
caffe.TEST)
#加载测试图像

image = caffe.io.load_image('../../data/fcn_label_full/fig/imag1_1-1135.jpg')
transpose_image = image.transpose(2,0,1) #把通道axis移到前面
transpose_image2 = transpose_image.reshape((1,)+transpose_image.shape) #增加一个axis匹配data的要求

#把图像拷贝到网络中
net.blobs['data'].reshape(1, # batch size
3, # 3-channel (BGR) images
50, 160) # image size is 50*160
net.blobs['data'].data[...] = transpose_image2

#执行网络前向运算
output = net.forward() #输出的是最后一层的top blob

#输出预测结果
#对每一个像素点输出其最大预测值对应的通道数
#因为我们知道shape(1,11,50,160)
temp_image = output['score1']
vote_image = np.zeros((50,160),dtype=np.int8)
for i in range(50):
for j in range(160):
max_index = 0
for ch in range(11):
if temp_image[0,ch,i,j] >= temp_image[0,max_index,i,j]:
max_index = ch
vote_image[i,j] = max_index
np.savetxt("vote_image.txt",vote_image,fmt="%02d")
===========================测试脚本复制 _end==================================

―

**Comment by jy9387** [comment]
另外，我对代码中的一些路径问题进行了一些微小的修改，以方便大家的使用，README中相应的部分也进行了调整，大家也可以update一下??


Jiang Yuan
School of Communication Engineering, Shanghai University, Shanghai, P.R. China


________________________________
发件人: geyijun <notifications@github.com>
发送时间: 2016年7月9日 11:30
收件人: SHUCV/digit
抄送: jy9387; Comment
主题: Re: [SHUCV/digit] 训练的时候不收敛 (#1)


谢谢作者的回复。我重新将问题简化并阐述一下，
我当前需求是:
按照项目中的readme把整个digit项目中的效果可以跑出来。
我详细记录我的操作过程如下:
步骤1: git clone digit 项目到本地。
步骤2: 下载data，解压到digit/data目录下 （unzip digit_data.zip）。

第一阶段:用 【作者训练好的网络】+【训练数据集中的图片】，验证效果。
步骤1>按照作者readme中的说，download pre-trained model 得：fcn11_full_iter_15000.caffemodel.放在digit/examples/fdigit/
步骤2>用python写一个简单脚本进行测试。见脚本my_test.py
步骤3>通过人工校验输出结果的正确性。发现完全不对(几乎95%是背景类，仅底边有一点点的0 和1 分类)。
我的问题:::为啥会这样？？？

说明1：caffe是7月1号安装的github上的版本。
说明2: 上述过程没有任何修改网络参数的动作，仅使用pre-trained model

===========================测试脚本复制 _begin==================================
#!/usr/bin/env python

-- coding:utf-8 --

from future import division
caffe_root = '/opt/caffe/'
import numpy as np
import sys,os
sys.path.insert(0, caffe_root + 'python')
import caffe

#创建网络
net = caffe.Net('./fcn11_deploy.prototxt',
'./fcn11_full_iter_15000.caffemodel',
caffe.TEST)
#加载测试图像

image = caffe.io.load_image('../../data/fcn_label_full/fig/imag1_1-1135.jpg')
transpose_image = image.transpose(2,0,1) #把通道axis移到前面
transpose_image2 = transpose_image.reshape((1,)+transpose_image.shape) #增加一个axis匹配data的要求

#把图像拷贝到网络中
net.blobs['data'].reshape(1, # batch size
3, # 3-channel (BGR) images
50, 160) # image size is 50*160
net.blobs['data'].data[...] = transpose_image2

#执行网络前向运算
output = net.forward() #输出的是最后一层的top blob

#输出预测结果
#对每一个像素点输出其最大预测值对应的通道数
#因为我们知道shape(1,11,50,160)
temp_image = output['score1']
vote_image = np.zeros((50,160),dtype=np.int8)
for i in range(50):
for j in range(160):
max_index = 0
for ch in range(11):
if temp_image[0,ch,i,j] >= temp_image[0,max_index,i,j]:
max_index = ch
vote_image[i,j] = max_index
np.savetxt("vote_image.txt",vote_image,fmt="%02d")
===========================测试脚本复制 _end==================================

―

**Comment by geyijun** [comment]
谢谢作者回复，我上面第一阶段遇到的问题，已经解决，我先自己解答一下。
原因1> 网络模型中对应的数据格式 和测试数据时输入的不一致。
   通过conver.py我们知道训练时输入的图像的值【0-255】。
   而我测试脚本中输入图像数据的值范围【0-1】 导致测试结果异常。
原因2> fcn11_deploy.prototxt 文件中的input_dim: 500值设置的不合理。
   导致数据内容完全不对。
上面两个问题修改之后，基本上第一阶段的目标是达到了。至此第一阶段的疑问结束。

那么我现在就是进入第二阶段用 【作者的数据（图像+标签）】+重新训练，验证训练后的模型的效果。
这个阶段遇到的问题，我重新阐述一下。

作者> *如果想按照README自己训练model，还请使用我们的caffe，因为代码跟官方的还是有一些不一  
          样的，主要是对softmax_loss.cpp层进行了一些修改，加入了多分类时正负样本的均衡，从而改善
          训练过程。之前你说过loss到0.6无法继续下降，有可能是因为这个原因。
问题: 能麻烦您再详细讲解一下 “softmax_loss_layer.cpp中加入了多分类时正负样本的均衡”的具体位置吗？？？ 因为我粗略对比了softmax_loss_layer.cpp和caffe 官网上下的，似乎逻辑上，没有什么大的差异。所以，还是想恳请您，再再详细讲解一下 “softmax_loss_layer.cpp中的改动。谢谢。

作者> 我对git上的代码做过测试，可以正常编译
问题: 确实是可以编译，但是编译出来的内容中少了一个_caffe.so库（ 这个是python api的封装），导致python调用异常，也就是说，直接用您提供的caffe环境是无法在python中调用，这个自然就无法执行conveer,py ，solve.py的。不知道我是不是哪里搞错了。

**Comment by geyijun** [comment]
@jy9387

**Comment by jy9387** [comment]
请问你在make caffe之后，有没有make pycaffe
在 2016年7月9日，18:15，geyijun <notifications@github.com<mailto:notifications@github.com>> 写道：

中

**Comment by geyijun** [comment]
谢谢提醒，确实是漏掉了make pycaffe。后面都OK了。

**Comment by geyijun** [comment]
非常感谢，确认按照Readme的做法，验证没有问题。

**Comment by jy9387** [comment]
不客气


Jiang Yuan
School of Communication Engineering, Shanghai University, Shanghai, P.R. China


________________________________
发件人: geyijun <notifications@github.com>
发送时间: 2016年7月12日 19:19
收件人: SHUCV/digit
抄送: jy9387; Mention
主题: Re: [SHUCV/digit] 训练的时候不收敛 (#1)


非常感谢，确认按照Readme的做法，验证没有问题。

―

**Comment by Darknesszlx** [comment]
“如果想按照README自己训练model，还请使用我们的caffe，因为代码跟官方的还是有一些不一样的，主要是对softmax_loss.cpp层进行了一些修改，加入了多分类时正负样本的均衡，从而改善训练过程。”
请问一下修改的位置在哪？

**Comment by jy9387** [comment]
你好，这部分的代码量不大，如果感兴趣可以参考caffe官方版本的源文件对比一下就知道了。


Jiang Yuan
School of Communication Engineering, Shanghai University, Shanghai, P.R. China


________________________________
发件人: Darknesszlx <notifications@github.com>
发送时间: 2016年7月25日 18:39
收件人: SHUCV/digit
抄送: Yuan.Jiang; Mention
主题: Re: [SHUCV/digit] 训练的时候不收敛 (#1)


“如果想按照README自己训练model，还请使用我们的caffe，因为代码跟官方的还是有一些不一样的，主要是对softmax_loss.cpp层进行了一些修改，加入了多分类时正负样本的均衡，从而改善训练过程。”
请问一下修改的位置在哪？

―

**Comment by Darknesszlx** [comment]
首先感谢回复，我对比了一下，发现softmax_loss.cpp, softmax_loss.cu,与caffe官方版本的源文件并没有差别。

**Comment by Darknesszlx** [comment]
首先感谢回复，我对比了一下，发现softmax_loss_layer.cpp, softmax_loss_layer.cu, softmax_layer.cpp, softmax_layer.cu这四个文件与caffe官方版本的源文件并没有差别。

**Comment by jy9387** [comment]
对不起，我也发现了这一点，这应该是某次push时的错误，最近放假在家，我会尽快修复，非常抱歉，如果需要看下我们改后的softmax_loss_layer.cpp，可以参考这里：https://github.com/jy9387/DeepSkeleton/blob/master/src/caffe/layers/softmax_loss_layer.cpp

[https://avatars1.githubusercontent.com/u/14325647?v=3&s=400]<https://github.com/jy9387/DeepSkeleton/blob/master/src/caffe/layers/softmax_loss_layer.cpp>

jy9387/DeepSkeleton<https://github.com/jy9387/DeepSkeleton/blob/master/src/caffe/layers/softmax_loss_layer.cpp>
github.com
DeepSkeleton - :racehorse:code for cvpr2016 paper "Object Skeleton Extraction in Natural Images by Fusing Scale-associated Deep Side Outputs"




Jiang Yuan
School of Communication Engineering, Shanghai University, Shanghai, P.R. China


________________________________
发件人: Darknesszlx <notifications@github.com>
发送时间: 2016年7月26日 9:44
收件人: SHUCV/digit
抄送: Yuan.Jiang; Mention
主题: Re: [SHUCV/digit] 训练的时候不收敛 (#1)


首先感谢回复，我对比了一下，发现softmax_loss_layer.cpp, softmax_loss_layer.cu, softmax_layer.cpp, softmax_layer.cu这四个文件与caffe官方版本的源文件并没有差别。

―

**Comment by Darknesszlx** [comment]
好的，非常感谢！

**Comment by Darknesszlx** [comment]
上面的链接里实现的是正负两类的情况，loss与样本个数存在一个“反比”的关系。
pos_loss = pos_loss * count_neg / (count_pos + count_neg);
neg_loss = neg_loss * count_pos / (count_pos + count_neg);
 loss = pos_loss + neg_loss;
请问一下，多个label时这段代码应该怎么改？
